[msan] Re-exec with no ASLR if memory layout is incompatible on Linux

thurstond opened the pull request:
This ports the change from TSan (https://github.com/llvm/llvm-project/commit/0784b1eefa36d4acbb0dacd2d18796e26313b6c5).

A key difference is that TSan initializes the allocator prior to
CheckAndProtect, while MSan initializes the allocator afterwards;
this slightly simplifies the MSan patch. Nonetheless, we need to check
that the allocator layout is compatible with ASLR. This must be done silently ("dry run") because any warning messages may cause tests to fail.

Testing notes: run 'sudo sysctl vm.mmap_rnd_bits=32; ninja check-msan'
before and after this patch.

N.B. aggressive ASLR may also cause the app to overlap with the allocator region; this was fixed in https://github.com/llvm/llvm-project/commit/af2bf86a372cacf5f536bae06e2f2d3886eefb7b

thurstond committed:
[msan] Re-exec with no ASLR if memory layout is incompatible on Linux

This ports the change from TSan (https://github.com/llvm/llvm-project/commit/0784b1eefa36d4acbb0dacd2d18796e26313b6c5).

A key difference is that TSan initializes the allocator prior to
CheckAndProtect, while MSan initializes the allocator afterwards;
this slightly simplifies the MSan patch. Nonetheless, we need to check
that the allocator layout is compatible with ASLR. Since the information is
not readily available in msan.h, we duplicate the information from
msan_allocator.cpp, and create a new MappingDesc::ALLOCATOR type.

Testing notes: run 'sudo sysctl vm.mmap_rnd_bits=32; ninja check-msan'
before and after this patch.

vitalybuka commented on compiler-rt/lib/msan/msan.h:
```diff
@@ -106,7 +107,8 @@ const MappingDesc kMemoryLayout[] = {
     {0x600000000000ULL, 0x610000000000ULL, MappingDesc::ORIGIN, "origin-1"},
     {0x610000000000ULL, 0x700000000000ULL, MappingDesc::INVALID, "invalid"},
-    {0x700000000000ULL, 0x800000000000ULL, MappingDesc::APP, "app-3"}};
+    {0x700000000000ULL, 0x740000000000ULL, MappingDesc::ALLOCATOR, "allocator"},
```
why do we need another enum value?

thurstond replied:
InitShadow is called before the allocator is initialized, so we need a way to tell InitShadow that we want to check that the allocator memory range is available (but not map it, because the allocator will do that).

The existing enum values don't suffice:
- MappingDesc::APP does not call CheckMemoryRangeAvailability()
- MappingDesc::INVALID does call CheckMemoryRangeAvailability(), but it will also be protected, which we don't want
- MappingDesc::SHADOW also calls CheckMemoryRangeAvailability(), but it will map the memory, which we also don't want
- MappingDesc::ALLOCATOR will call CheckMemoryRangeAvailability() but will not map or protect the memory

thurstond replied:
Currently, the kernel might randomly map the app into the same region that the allocator expects (this isn't a problem with low-entropy ASLR, because it never maps into the bottom part of the app region, which is where the allocator typically lives). We therefore need to enforce that the APP and ALLOCATOR regions are disjoint.

vitalybuka replied:
Would it possible to introduce MappingDesc::ALLOCATOR in a separate patch?

thurstond replied:
Do you mean introduce MappingDesc::ALLOCATOR as an NFC patch?

vitalybuka replied:
yes, and related conditions

thurstond replied:
Ok, will do.

thurstond replied:
Created https://github.com/llvm/llvm-project/pull/85153

vitalybuka commented on compiler-rt/lib/msan/msan_linux.cpp:
```diff
@@ -138,27 +144,64 @@ bool InitShadow(bool init_origins) {
+  // Start with dry run: check layout is ok, but don't print warnings because
+  // warning messages will cause tests to fail (even if we successfully re-exec
+  // after the warning).
+  bool success = InitShadow(__msan_get_track_origins(), true);
```
How expensive doing InitShadow twice?

thurstond replied:
When InitShadow is called the first time (dry run), it calls CheckMemoryRangeAvailability() but not MmapFixedSuperNoReserve() or ProtectMemoryRange(), so it shouldn't be that expensive.

vitalybuka replied:
It traverse `MemoryMappingLayout proc_maps(/*cache_enabled*/true);`

can we instead if `!dry_run` skip `CheckMemoryRangeAvailability`

thurstond replied:
Done, thanks for the feedback!

thurstond committed:
Skip CheckMemoryRangeAvailability if !dry_run, as suggested by Vitaly

thurstond committed:
Bug fix: if the memory layout is incompatible and we did not re-exec, we must
still return false. (The second InitShadow call does not perform the
same checks that the first InitShadow call did.)

vitalybuka commented on compiler-rt/lib/msan/msan_linux.cpp:
```diff
@@ -197,7 +197,7 @@ bool InitShadowWithReExec(bool init_origins) {
   // The earlier dry run didn't actually map or protect anything. Run again in
   // non-dry run mode.
-  success = InitShadow(__msan_get_track_origins(), false);
+  success = success && InitShadow(__msan_get_track_origins(), false);
```
return success && InitShadow(__msan_get_track_origins(), false);

thurstond replied:
Done!

thurstond committed:
Simplify code per Vitaly's comment

aeubanks commented:
We're seeing similar issues with dfsan, e.g. https://ci.chromium.org/ui/p/chromium/builders/try/linux_upload_clang/5066/overview

```
 FAIL: DataFlowSanitizer-x86_64 :: lookup_table.c (22061 of 81767)
 ******************** TEST 'DataFlowSanitizer-x86_64 :: lookup_table.c' FAILED ********************
 Exit Code: 139
 
 Command Output (stderr):
 --
 RUN: at line 1: /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/./bin/clang  -fsanitize=dataflow   -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta   /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/test/dfsan/lookup_table.c -mllvm -dfsan-combine-offset-labels-on-gep=false -mllvm -dfsan-combine-pointer-labels-on-load=false -mllvm -dfsan-combine-taint-lookup-table=remap_to_upper -DLOOKUP_TABLE -o /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/dfsan/X86_64Config/Output/lookup_table.c.tmp &&  /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/dfsan/X86_64Config/Output/lookup_table.c.tmp
 + /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/./bin/clang -fsanitize=dataflow -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/test/dfsan/lookup_table.c -mllvm -dfsan-combine-offset-labels-on-gep=false -mllvm -dfsan-combine-pointer-labels-on-load=false -mllvm -dfsan-combine-taint-lookup-table=remap_to_upper -DLOOKUP_TABLE -o /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/dfsan/X86_64Config/Output/lookup_table.c.tmp
 + /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/dfsan/X86_64Config/Output/lookup_table.c.tmp
 FATAL: Code 0x618710e92e60 is out of application range. Non-PIE build?
 /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/dfsan/X86_64Config/Output/lookup_table.c.script: line 3: 850288 Segmentation fault      (core dumped) /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/dfsan/X86_64Config/Output/lookup_table.c.tmp
```

thurstond commented:
> We're seeing similar issues with dfsan, e.g. https://ci.chromium.org/ui/p/chromium/builders/try/linux_upload_clang/5066/overview
> 
> ```
>  FAIL: DataFlowSanitizer-x86_64 :: lookup_table.c (22061 of 81767)
>  ******************** TEST 'DataFlowSanitizer-x86_64 :: lookup_table.c' FAILED ********************
>  Exit Code: 139
>  
>  Command Output (stderr):
>  --
>  RUN: at line 1: /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/./bin/clang  -fsanitize=dataflow   -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta   /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/test/dfsan/lookup_table.c -mllvm -dfsan-combine-offset-labels-on-gep=false -mllvm -dfsan-combine-pointer-labels-on-load=false -mllvm -dfsan-combine-taint-lookup-table=remap_to_upper -DLOOKUP_TABLE -o /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/dfsan/X86_64Config/Output/lookup_table.c.tmp &&  /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/dfsan/X86_64Config/Output/lookup_table.c.tmp
>  + /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/./bin/clang -fsanitize=dataflow -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/test/dfsan/lookup_table.c -mllvm -dfsan-combine-offset-labels-on-gep=false -mllvm -dfsan-combine-pointer-labels-on-load=false -mllvm -dfsan-combine-taint-lookup-table=remap_to_upper -DLOOKUP_TABLE -o /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/dfsan/X86_64Config/Output/lookup_table.c.tmp
>  + /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/dfsan/X86_64Config/Output/lookup_table.c.tmp
>  FATAL: Code 0x618710e92e60 is out of application range. Non-PIE build?
>  /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/dfsan/X86_64Config/Output/lookup_table.c.script: line 3: 850288 Segmentation fault      (core dumped) /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/dfsan/X86_64Config/Output/lookup_table.c.tmp
> ```

Thank you Arthur for the heads up! I've drafted a pull request to port this patch to DFSan (https://github.com/llvm/llvm-project/pull/85674)
